### PR TITLE
NetBSD support with envsys ioctl and plist

### PR DESCRIPTION
**Pull request by naguam**
New examples, simple.rs only refreshed the first battery, and this does not help checking for iterator validity for development, all.rs solves the problem. The example oneshot.rs is just for comfort avoiding loop.

NetBSD exposes battery information through one ioctl so for now we have to "waste" resource getting all sensors info for each single battery refresh. We can't assume the library user will want to refresh all bateries at the same time even if 90% of time it will probably be.

The result of the ioctl is a plist (apple format) that needs to be parsed and accessed.

The plist library only uses options, so this is the reason for utils.rs. These are all oneline functions to reduce the lenght of chains in the netbsd platform device.rs file.

NetBSD does not expose battery vendor, model, technology, cycle among other things for now, but these are optional.

Got inspiration from the FreeBSD port regarding having a better understanding on how the library works.

Thanks to unitedbsd.com community for the guidance, and shout out to https://github.com/distatus/battery/blob/master/battery_netbsd.go as well as the NetBSD devs for the great man pages and the clear code allowing to read what happens for undocumented things.

My code has comments for anyone interested to understand what happens.

By lack of hardware (only tested on one laptop), I couldn't test some specific things such as cases where battery use the watt units instead of ampere. I implemented that with the NetBSD man pages and inspiration from the FreeBSD port.

This port was only tested with a battery using the acpibat driver, but this should work with any battery that follow the same standard in the the sysmon_envsys format and fields.

### CODE REVIEW

**Comment by naguam** [comment]
This repository not having ffi anymore, I think the README could be updated.

**Comment by naguam** [comment]
I might need help with setting the build CI for NetBSD though. The NetBSD vm-action from the same source as FreeBSD seems to lack libexecinfo.

File: examples/all.rs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+extern crate starship_battery as battery;
+
+// use std::io;
```
Comment (ThalusA) [inline examples/all.rs:3]:
```suggestion
```

**Review by ThalusA — APPROVED** [review]

**Comment by davidkna** [comment]
@naguam If this only requires a newer build image, maybe you can point cross to a more recent one via `Cross.toml`?

Furthermore, you think the `plist`-helpers you be replaced by instead using a struct with `derive(serde::Deserialize)`?

**Comment by naguam** [comment]
> @naguam If this only requires a newer build image, maybe you can point cross to a more recent one via `Cross.toml`?
> 
> Furthermore, you think the `plist`-helpers you be replaced by instead using a struct with `derive(serde::Deserialize)`?

I've never used Serde directly, It might me possible to use it, but only might. A friend (Actually @ThalusA) told me it could be possible to use it, but on the other hand, I don't think I can rely on the order of the data and format.

This is why these are property lists and not plain C structs exposed throughout the ioctl after all like FreeBSD does in acpi.rs after all.

**Comment by naguam** [comment]
@davidkna sorry for all the edits in my answers, I just tried to make it better.

**Comment by davidkna** [comment]
> I don't think I can rely on the order of the data and format, it could be if we read how they are built in kernel code, but not according to userspace documentation AFAIK.

`serde` should not care about the ordering in the `plist`. Nevertheless, I am not sure if the way the library maps to `serde` would not cause any issues.

To elaborate on how the `Cross.toml` file could help: `Cross.toml` can be used to configure the use of different docker base images for each target, as I mentioned [a more recent one](https://github.com/orgs/cross-rs/packages) could solve the compile issues.

**Comment by naguam** [comment]
I think I don't understand.

Of course cross can specify other ways of building, but the ways of building and versions it recognises depends on the versions of cross as far as i can understand.

The way it is setup here, 
https://github.com/starship/rust-battery/blob/main/.github/workflows/ci.yml
is installed by https://github.com/taiki-e/install-action which if we check.
https://github.com/taiki-e/install-action/blob/main/manifests/cross.json

image

Gets its version from the release page tarball.
https://github.com/cross-rs/cross/releases

Which means that the available target either Dockerfile image or direct script, are not containing the fix.

Maybe I missed something, but I probably need guidance.

Should I push in the repo the even more recent netbsd edge script, and try running it manually.

I don't think it is maintainable. Maybe my first proposal is easier (disabling the backtrace until new release comes)

**Comment by davidkna** [comment]
@naguam `Cross.toml` configures `cross` itself and is unrelated to the version that is installed ([reference](https://github.com/cross-rs/cross/blob/main/docs/config_file.md)). It might be possible to apply the required fix-ups via this file too via `pre-build`.

**Comment by naguam** [comment]
Ok I think I managed to fix the CI by manually implementing the fix.

Also added clippy fixes.

Once the next proper release of the cross-rs will be published this hot fix will have to be removed (especially if NetBSD version is bumped to 10).

**Comment by naguam** [comment]
> By lack of hardware (only tested on one laptop), I couldn't test some specific things such as cases where battery use the watt units instead of ampere. I implemented that with the NetBSD man pages and inspiration from the FreeBSD port.

Virtualbox translates an acpi battery which is in watt unit.

I have tested my code and my assumptions were correct, so it works, did not need to make changes.

**Review by davidkna — COMMENTED** [review]
Confirmed working on real HW. LGTM, apart from two small additional suggestions.

**Comment by naguam** [comment]
For the CI thanks, I did not know I could do that

(it is probably in the documentation of cross-rs though)

However don't merge it too quickly

c.f. https://www.unitedbsd.com/d/1311-battery-data-such-as-vendor-model/22

It works but it is based on descriptions, this need to do a full rework, probably using serde directly.

I don't have the motivation for now.

**Comment by davidkna** [comment]
In that case, perhaps return an error if a different driver is used for now?

**Comment by naguam** [comment]
> In that case, perhaps return an error if a different driver is used for now?

Yes with the new patch only acpibat is supported I I do not rely on descriptions anymore.

**Review by davidkna — APPROVED** [review]

**Comment by davidkna** [comment]
Thanks for the contribution @naguam!

**Pull request merged**